webapp/src/app/shared/widgets/npc-states-widget/npc-states/npc-states.component.ts

```diff
@@ -125,4 +138,30 @@ export class NpcStatesComponent implements OnInit {
 	cancel() {
 		this.exit.error('cancel');
 	}
+	
+	updateEventWarnings(updatedEvents: EventType[]) {
+		function hasEventOfTypeRecursive(object: any, type: string): boolean {
+			if (typeof object !== 'object') {
+				return false;
+			}
+			for (const key of Object.keys(object)) {
+				if ((key === 'type' && object[key] === type) || hasEventOfTypeRecursive(object[key], type)) {
+					return true;
+				}
+			}
+			return false;
+		}
+		
+		this.missingTradeEvent =
+			updatedEvents.length > 0 &&
+			!hasEventOfTypeRecursive(updatedEvents, 'START_NPC_TRADE_MENU');
+	}
+	
+	get isTradeEvent() {
+		return this.eventType === EventArrayType.Trade;
+	}
+	
+	get eventTypes(): string[] {
```

Comment:
should be calculated only once and reused. functions in templates are really inefficient